### PR TITLE
alternative to rlist

### PR DESCRIPTION
**Pull request by condret**
random access to element with certain priority in O(1)
random access to element with certain id in O(1)

### CODE REVIEW

**Comment by radare** [comment]
I think we cna find a better name for this long snake case thing

**Comment by condret** [comment]
what snake case?

**Comment by radare** [comment]
r_oid_storage_get_id

File: libr/util/idpool.c [hunk 1]
```diff
@@ -195,3 +196,210 @@ R_API void r_id_storage_free(RIDStorage* storage) {
...
+		_id = (ut32 *)storage->permutation->data[od];
+		if (_id && (id == *_id)) {
+			break;
+		}
+	}
+	if (od == (storage->permutation->pool->next_id - 1)) {
+		return true;
+	}
+	memmove (&storage->permutation->data[od], &storage->permutation->data[od + 1],
+		(storage->permutation->pool->next_id - (od + 1)) * sizeof(void *));
```
Comment (radare) [inline libr/util/idpool.c]: missing space before (

**Comment by radare** [comment]
name it r_oids_

**Comment by condret** [comment]
will rename later, first fix it for windows build

**Comment by condret** [comment]
will rename later, first fix it for windows build

**Comment by radare** [comment]
u dont use windows, why that matters? :D

**Comment by XVilka** [comment]
Many people use Windows.

**Comment by radare** [comment]
i fixed that windows problem. now you can rename

**Comment by condret** [comment]
not much time, but I do my best

File: libr/util/idpool.c (hunk 1 above)
Comment (condret) [reply libr/util/idpool.c]: wat? doesn't it look fine this way?

File: libr/util/idpool.c [hunk 2]
```diff
@@ -105,8 +110,8 @@ R_API bool r_id_storage_set(RIDStorage* storage, void* data, ut32 id) {
 		return false;
 	}
 	n = get_msb (id + 1);
-	if (n > (storage->size - (storage->size / 4))) {
-		if (n < (storage->pool->last_id / 2)) {
+	if (n > ((storage->size/2) + (storage->size / 4))) {
```
Comment (radare) [inline libr/util/idpool.c]: spaces around /

File: libr/util/idpool.c [hunk 3]
```diff
@@ -195,3 +200,263 @@ R_API void r_id_storage_free(RIDStorage* storage) {
...
+				return true;
+			}
+		}
+	}
+	return false;
+}
+
+R_API bool r_oids_add(ROIDStorage *storage, void *data, ut32 *id, ut32 *od) {
+	ut32 *_id;
+	if (!storage || !id || !od || !(_id = R_NEW (ut32))) {
```
Comment (radare) [inline libr/util/idpool.c]: R_NEW(ut32)? why not use a pool instead, thats unnefficient

File: libr/util/idpool.c [hunk 4]
```diff
@@ -195,3 +200,263 @@ R_API void r_id_storage_free(RIDStorage* storage) {
...
+	for (od = 0; od < storage->permutation->pool->next_id; od++) {
+		_id = (ut32 *)storage->permutation->data[od];
+		if (_id && (id == *_id)) {
+			break;
+		}
+	}
+	if (od == (storage->permutation->pool->next_id - 1)) {
+		return true;
+	}
+	memmove (&storage->permutation->data[od], &storage->permutation->data[od + 1],
```
Comment (radare) [inline libr/util/idpool.c]: this looks like an slist

File: libr/util/idpool.c [hunk 5]
```diff
@@ -195,3 +200,263 @@ R_API void r_id_storage_free(RIDStorage* storage) {
...
+			}
+		}
+	}
+	if ((id = storage->permutation->data[i])) {
+		return cb (user, storage->data->data[*id], *id);
+	}
+	return true;
+}
+
+#if __WINDOWS__ && !__CYGWIN__
```
Comment (radare) [inline libr/util/idpool.c]: wtf

File: libr/util/idpool.c [hunk 6]
```diff
@@ -195,3 +200,263 @@ R_API void r_id_storage_free(RIDStorage* storage) {
...
+	return st->cmp (r_id_storage_get (st->data, a_id),
+			r_id_storage_get (st->data, b_id));
+}
+
+R_API void r_oids_sort (ROIDStorage *storage, ROIDStorageCompareCb cmp) {
+	if (!storage || !storage->data || !storage->permutation) {
+		return;
+	}
+	storage->cmp = cmp;
+#if __WINDOWS__ && !__CYGWIN__
```
Comment (radare) [inline libr/util/idpool.c]: why dat?

File: libr/util/idpool.c (hunk 3 above)
Comment (condret) [reply libr/util/idpool.c]: bc this way we can use _id = NULL for error somewhere else in the code. also void * != ut32

File: libr/util/idpool.c (hunk 6 above)
Comment (condret) [reply libr/util/idpool.c]: qsort_r does not exist on windows, windows calls it qsort_s. qsort_s does the parameter order wrong

**Comment by radare** [comment]
Ut32 fits inside void*

**Comment by radare** [comment]
conflicts!

**Comment by radare** [comment]
please rebase and squash your commits. (and solve the conflict)

File: libr/util/idpool.c [hunk 7]
```diff
@@ -201,8 +228,226 @@ static bool _list(void* user, void* data, ut32 id) {
...
+		return false;
+	}
+	for (od = 0; od < storage->ptop; od++) {
+		if (id == storage->permutation[od]) {
+			break;
+		}
+	}
+	if (od == storage->ptop) {
+		return false;
+	} else if (od == 0) {
```
Comment (radare) [inline libr/util/idpool.c]: dont else after return

File: libr/util/idpool.c [hunk 8]
```diff
@@ -201,8 +228,226 @@ static bool _list(void* user, void* data, ut32 id) {
...
+		return true;
+	}
+	memmove (&storage->permutation[1], &storage->permutation[0], od * sizeof(ut32));
```
Comment (radare) [inline libr/util/idpool.c]: missing space before (

File: libr/util/idpool.c [hunk 9]
```diff
@@ -201,8 +228,226 @@ static bool _list(void* user, void* data, ut32 id) {
...
+}
+
+R_API void r_oids_delete (ROIDStorage *storage, ut32 id) {
+	if (!r_oids_to_front (storage, id)) {
+		return;
+	}
+	r_id_storage_delete (storage->data, id);
+	storage->ptop--;
+	if (!storage->ptop) {
+		R_FREE(storage->permutation);
```
Comment (radare) [inline libr/util/idpool.c]: missing space (

**Comment by radare** [comment]
its red and there are cornflakes. pls fix

**Comment by radare** [comment]
Green is the new red

File: libr/util/idpool.c (hunk 5 above)
Comment (radare) [reply libr/util/idpool.c]: And this?

File: libr/util/idpool.c (hunk 6 above)
Comment (ret2libc) [reply libr/util/idpool.c]: apart from the name, here it seems parameters order is the same.

**Comment by ret2libc** [comment]
@condret a couple of comments/questions:
1) can you update the PR title? i think this PR has nothing to do with rlist in general and it creates confusion
2) exactly, what's the name of the data structure you are implementing? is it a known data structure? where is it going to be used? why? do we really need a new ad-hoc data structure for something we don't even have yet?

File: libr/util/idpool.c [hunk 10]
```diff
@@ -201,8 +228,227 @@ static bool _list(void* user, void* data, ut32 id) {
...
+		return storage->data->data[storage->permutation[0]];
+	}
+	return NULL;
+}
+
+//return the element with highest order
+R_API void *r_oids_first (ROIDStorage *storage) {
+	if (storage && storage->data && storage->data->data
+		&& storage->permutation) {
+		return storage->data->data[storage->permutation[storage->ptop - 1]];
```
Comment (radare) [inline libr/util/idpool.c]: ptop will be always > 0 here?

**Comment by ret2libc** [comment]
@radare 

I see this has been merged... I'd have personally waited a bit more actually, since I'm not sure it is necessary at all. I was discussing with @condret that this could just be a rlist (for sorting things by priority or whatever) + an hashmap for fast access to elements by "id". 

I don't like the fact that this data structure is pretty much useful only for IOMap, it's not generic enough... for example, i don't see any reason for it to have a RIDStorage inside. Again, the same thing could have been done with 2 data structures, keeping a (conceptually) simpler code.

That's the problem of creating data structures *before* actually knowing what/where/which operations you need. For example, do you really need random access by priority in O(1)? And if we *really* do (which we can know only after having seen the real use case and measured stuff) why a separate hashmap by priority wasn't enough?

**Comment by radare** [comment]
Totally agree

**Pull request merged**